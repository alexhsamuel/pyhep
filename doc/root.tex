\section{\module{hep.root} -- Root interface}

\declaremodule{extension}{hep.root}
\modulesynopsis{Root interface}

The \module{hep.root} module provides access from Python to features in
Root, a framework for high energy physics data management and analysis.
The applicable components of Root are included with \pyhep and need not
be provided separately.

\subsection{Root files}

Root includes an implementation of histograms and ntuples (which are
called ``trees'' in Root).  It implements a file format for storing
these in structured disk files.  \pyhep provides read and write access
to histograms and trees stored in Root files.

The \function{open} function access a Root file:

\begin{funcdesc}{open}{path\optional{, mode="r", purge_cycles=1}}
 Create or open a Root file at \var{path}.  

 \var{mode} is the access mode, analogous to Python's built-in
 \function{file} and \function{open} functions.  If \code{"r"}, opens an
 existing file for reading.  If \code{"r+"}, \code{"w"}, \code{"a"}, or
 \code{"a+"}, opens an existing file for reading and writing, or creates
 the file if it doesn't exist.  If \code{"w+"}, creates a file or
 overwrites an existing file.

 If \var{purge_cycles} is set to a true value and the file is open for
 writing, the entire file will be purged when it is closed.  
\end{funcdesc}

The return value from \function{open} is a \class{File} object.  The
file remains open until it is deleted.  

In \pyhep, all paths in Root files are specified relative to the root
directory of the Root file.  The root directory itself is represented by
an empty string, and the path separator is a forward slash.  Note that
there is no notion of working directory or relative paths.

Root files also support revision cycles; see the Root documentation for
details.  Unless \function{root.open} is invoked with
\code{purge_cycles=0} (or the file is opened read-only), \pyhep
``purges'' the entire file when the file is closed.  This removes cycles
except the most recent for each object in the file, whether or not the
object was written or revised in the current session.

\pyhep stores additional information along with tables and histograms in
Root files.  This information, called \emph{metadata}, consists of
useful information about histograms and tables that is not handled
directly by the Root file format.  \pyhep stores metadata in
subdirectories named ``_pyhep_metadata'' in the Root file.  These
subdirectories and their contents may be removed safely without injuring
the primary data in the histograms and tables they are associated with,
but doing so will delete configuration and annotations accessible from
\pyhep.

\subsubsection{\class{File} objects}

A \class{File} object has the following methods and attributes:

\begin{memberdesc}{writeable}
 True if the file is open for writing as well as reading.
\end{memberdesc}

\begin{methoddesc}{listdir}{\optional{path}}
 Return the contents of the directory specified by \var{path}.  The
 return value is a sequence of \code{(name, title, class_name)} tuples,
 where \var{name} is the object's name, \var{title} is the object's
 title, and \var{class_name} is the object's class name.
\end{methoddesc}

\begin{methoddesc}{mkdir}{path\optional{, title}}
 Create a new directory at \var{path}.  Optionally, a title may be
 specified in the \var{title} argument.
\end{methoddesc}

\begin{methoddesc}{rmdir}{path}
 Remove the directory at \var{path}.
\end{methoddesc}

\begin{methoddesc}{save}{path, object}
 Store a histogram \var{object} at \var{path}.  One- and two-dimensional
 histograms are supported.  
\end{methoddesc}

\begin{methoddesc}{load}{path}
 Load a histogram or connect to an ntuple at \var{path}.  If \var{path}
 is a one- or two-dimensional histogram, returns a histogram object with
 its contents.  If \var{path} is a tree, returns a table object (see
 \module{hep.table}) accessing it.
\end{methoddesc}

\begin{methoddesc}{createTable}{path, schema\optional{, title,
separate_branches=0, branch_name="default"}}
 Create a table as a tree at \var{path} in the Root file, using
 columns from \var{schema}, a \class{hep.table.Schema} instance.  

 By default, column values is stored in a leaves grouped in a single
 branch, named "default".  Another name may be specified for the branch
 with the \var{branch_name} argument.  Alternately, if
 \var{separate_branches} is true, each leaf is placed in a separate
 branch, whose name is the column name.

 These column types are supported: \constant{"int8"},
 \constant{"int16"}, \constant{"int32"}, \constant{"float32"}, and
 \constant{"float64"}.
\end{methoddesc}

\begin{methoddesc}{remove}{path}
 Remove the tree or histogram at \var{path}.  Use \method{rmdir} to
 remove a directory.
\end{methoddesc}

\begin{methoddesc}{normpath}{path}
 Return the canonical form of absolute path \var{path} in the Root
 file.
\end{methoddesc}

\begin{methoddesc}{join}{*components}
 Join together \var{components} to form an absolute path.
\end{methoddesc}

\begin{methoddesc}{split}{path}
 Return \code{(dir, name)}, where \var{dir} is the portion of \var{path}
 up to the last directory separator, and \var{name} is the remainder.
\end{methoddesc}

