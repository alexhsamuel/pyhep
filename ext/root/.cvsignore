root.py
root_wrap.cc
root_wrap.doc
*Dict.cc
*Dict.h
*.pyc
depend.mk
ROOTSYS
