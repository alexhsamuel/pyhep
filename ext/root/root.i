//------------------------------------------------------------*- c++ -*-
//
// root.i
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

// SWIG interface file for Root libraries.

// -*- c++ -*-

%module root
%{
#include "TROOT.h"
#include "TKey.h"
#include "TDirectory.h"
#include "TFile.h"
#include "TLeaf.h"
#include "TLeafD.h"
#include "TLeafF.h"
#include "TLeafI.h"
#include "TList.h"
#include "TObject.h"
#include "TTree.h"
#include "TH1.h"
#include "TH2.h"
#include "TH3.h"
#include "TAxis.h"

class IndexError 
{
public:
  IndexError(int index) : index_(index) {};
  const int index_;
};

class StopIteration {};

%}

%inline %{
long 
getPointerValue(void* ptr) 
{
   return (long) ptr;
}
%}

%except(python) {
  try {
    $function
  }
  catch (IndexError e) {
    PyErr_Format(PyExc_IndexError, "%d", e.index_);
    return NULL;
  }
  catch (StopIteration) {
    PyErr_SetString(PyExc_StopIteration, "end of iterator");
    return NULL;
  }
}


class TObject
{
public:

  bool IsZombie() const;
  virtual const char* ClassName() const;
  virtual const char* GetName() const;
  virtual const char* GetTitle() const;
  virtual bool InheritsFrom(const char *classname) const;
  virtual void Print(const char* option) const;
  virtual int Write(const char* name=0, int option=0, int bufsize=0);

};


class TNamed
  : public TObject
{
public:

  virtual void SetName(const char* name);
  virtual void SetTitle(const char* title);

};


%pragma no_default

class TIterator
  : public TObject
{
public:

  virtual TObject* Next();

};

%addmethods TIterator {
  TIterator* __iter__() 
  {
    return self;
  };


  TObject* next()
  {
    TObject* object = self->Next();
    if (object == NULL)
      throw StopIteration();
    else
      return object;
  };
};



class TClass
  : public TObject
{
public:

  bool InheritsFrom(const char* cl) const;

};


class TKey
  : public TNamed
{
public:

  const char* GetClassName() const;
  short GetCycle() const;
  virtual TClass* IsA() const;

};


class TDirectory
  : public TNamed
{
public:

  virtual bool cd();
  virtual void Delete(const char* namecycle);
  virtual TObject* Get(const char* namecycle);
  TFile* GetFile() const;
  TKey* GetKey(const char* name, short cycle=9999);
  TList* GetList() const;
  TList* GetListOfKeys() const;
  TObject* GetMother() const;
  const char* GetPath() const;
  virtual TDirectory* mkdir(const char* name, const char* title);
  void Purge(short nKeep);
  
};


class TFile
  : public TDirectory
{
public:

  TFile(const char* fname, 
	char* option="", 
	const char* ftitle="", 
	int compress=1);

  virtual bool cd(const char *path = 0);
  virtual void Close(char* option="");
  virtual int Write(const char* name=0, int opt=0, int bufsiz=0);

};


class TList
  : public TObject
{
public:

  virtual int GetSize() const;
  virtual TObject* At(int idx) const;

};


%addmethods TList {
  int __len__() 
  {
    return self->GetSize();
  };


  TObject* __getitem__(int idx) 
  {
    if (idx < 0 || idx >= self->GetSize()) 
      throw IndexError(idx);
    else
      return self->At(idx);
  };
};



class TTree
  : public TObject
{
public:

  virtual int GetEntry(int entry=0, int getall=0);
  virtual double GetEntries() const;
  virtual TIterator* GetIteratorOnAllLeaves();
  virtual void Print();

};


class TLeaf
  : public TObject
{
public:
  
  virtual void SetAddress(void* add=0);

};

%addmethods TLeaf {
  void SetAddressInt(int add=0) 
  {
    self->SetAddress((void*) add);
  };
};


class TLeafD
  : public TLeaf
{
};


class TLeafF
  : public TLeaf
{
};


class TLeafI
  : public TLeaf
{
};


class TH1
  : public TNamed
{
public:

  virtual TAxis* GetXaxis();
  virtual TAxis* GetYaxis();
  virtual TAxis* GetZaxis();
  virtual double GetBinContent(int bin);
  virtual double GetBinError(int bin);
  virtual double GetEntries();
  virtual void SetBinContent(int bin, double content);
  virtual void SetBinError(int bin, double error);
  virtual void SetEntries(double entries);

};


class TH1F 
  : public TH1
{
public:

  TH1F(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1D
  : public TH1
{
public:

  TH1D(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1C
  : public TH1
{
public:

  TH1C(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1S
  : public TH1
{
public:

  TH1S(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH2
  : public TH1
{
public:

  virtual double GetBinContent(int binx, int biny);
  virtual double GetBinError(int binx, int biny);
  virtual void SetBinContent(int binx, int biny, double content);
  virtual void SetBinError(int binx, int biny, double error);

};


class TH2C
  : public TH2
{
public:

  TH2C(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2D
  : public TH2
{
public:

  TH2D(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2F
  : public TH2
{
public:

  TH2F(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2S
  : public TH2
{
public:

  TH2S(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH3
  : public TH1
{
};


class TAxis
  : public TObject
{
public:

  int GetNbins();
  double GetXmin();
  double GetXmax();

};
