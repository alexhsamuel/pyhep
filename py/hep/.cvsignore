*.pyc
*.pyo
config.py
*-install
pdt.pickle
