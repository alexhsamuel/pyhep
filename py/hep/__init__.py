#-----------------------------------------------------------------------
#
# module hep
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Module for general-purpose code."""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

import os
import signal
import traceback

#-----------------------------------------------------------------------
# exceptions
#-----------------------------------------------------------------------

class TimeoutException(Exception):
    """A timeout occurred."""

    pass



class NotSupportedError(RuntimeError):
    """A feature is not supported."""

    pass



class SyntaxError(Exception):

    pass



#-----------------------------------------------------------------------
# classes
#-----------------------------------------------------------------------

class Token:

    pass



class Timeout:

    def __timeout_handler(self, signal_number, frame):
        assert signal_number == signal.SIGALRM
        signal.signal(signal.SIGALRM, self.__old_handler)
        self.__active = 0
        raise TimeoutException, self.__message


    def __init__(self, time, message):
        self.__message = message
        self.__old_handler = \
            signal.signal(signal.SIGALRM, self.__timeout_handler)
        self.__active = 1
        signal.alarm(time)


    def cancel(self):
        if not self.__active:
            return
        signal.alarm(0)
        signal.signal(signal.SIGALRM, signal.SIG_DFL)
        self.__active = 0



class MappingAsObject:
    """Implement the object attribute protocol using a mapping.

    A 'MappingAsObject' wraps an object satifying the mapping protocol.
    Attribute references are converted to item gets and sets in the
    underlying mapping."""

    def __init__(self, mapping):
        """Wrap 'mapping' in the object procotol."""
        
        self.__dict__["_MappingAsObject__mapping"] = mapping


    _mapping = property(lambda self: self.__mapping)
    """Read-only access to the underlying mapping."""


    def __getattr__(self, name):
        try:
            return self.__mapping[name]
        except KeyError:
            raise AttributeError, name


    def __setattr__(self, name, value):
        try:
            self.__mapping[name] = value
        except KeyError, exception:
            raise AttributeError, str(exception)



class NrBreitWigner:
    """A non-resonant Breit-Wigner line shape amplitude factor."""

    def __init__(self, mass, width):
        """Construct a non-relativistic Breit-Wigner function.

        'mass' -- The resonance mass.
        
        'width' -- The resonance width."""

        self.__mass = mass
        self.__width = width


    def __call__(self, s):
        """Return the amplitude at energy-squared 's'."""

        mass = self.__mass
        return 1 / (s - complex(mass * mass, mass * self.__width))



#-----------------------------------------------------------------------
# functions
#-----------------------------------------------------------------------

try:
    enumerate = __builtins__.enumerate
except AttributeError:
    def enumerate(iterable):
        count = 0
        for item in iter(iterable):
            yield count, item
            count += 1


try:
    sorted = __builtins__.sorted
except AttributeError:
    def sorted(iterable, cmp=cmp):
        result = list(iterable)
        result.sort(cmp)
        return result


try:
    reversed = __builtins__.reversed
except AttributeError:
    def reversed(iterable):
        result = list(iterable)
        result.reverse()
        return result
    

def sort(sequence, sort_fn=None):
    """Return a copy of 'sequence', sorted by 'sort_fn'.

    'sort_fn' -- The comparison function to use.  If 'None', the default
    (built-in 'cmp') is used."""

    sequence = list(sequence)
    if sort_fn is None:
        sort_fn = cmp
    sequence.sort(sort_fn)
    return sequence


def binarySearch(sequence, value, compare=cmp):
    """Find 'value' in sorted 'sequence' by binary search.

    'sequence' -- A sequence of values.  It is assumed to be sorted by
    'compare'.

    'compare' -- A comparison function.

    returns -- The index of the largest element in 'sequence' smaller
    than 'value', or -1 if there is no such element."""

    if len(sequence) == 0:
        return -1
    if compare(value, sequence[0]) < 0:
        return -1
    if compare(value, sequence[-1]) > 0:
        return len(squence) - 1

    i0 = 0
    i2 = len(sequence) - 1
    while True:
        if i2 <= i0 + 1:
            return i2 - 1
        i1 = (i0 + i2) // 2
        if compare(value, sequence[i1]) < 0:
            i2 = i1
        else:
            i0 = i1


def mapFromKeys(key_sequence, value_function):
    """Construct a map from a sequence of keys.

    A map is created whose keys are the elements of 'key_sequence'.  For
    each key, the corresponding value is constructed from the key using
    'value_function'.  If 'sequence' contains a particular key more than
    once, only one is placed in the map (which one is undefined).

    'key_sequence' -- A sequence of key values.

    'value_function' -- A callable for determining the value for each
    element of the sequence.  It is invoked for each key, passing the
    kay as the single argument."""
    
    result = {}
    for key in key_sequence:
        result[key] = value_function(key)
    return result


def mapFromValues(key_function, value_sequence):
    """Construct a map from a sequence of values.

    A map is created whose values are the elements of 'value_sequence'.
    For each value, the corresponding key is constructed from the value
    using 'key_function'.  If 'key_function' assigns the same key to
    more than one element of the sequence, only one element is actually
    used in the map (which one is undefined).

    'key_function' -- A callable for determining the key for each
    element of the sequence.  It is invoked for each value, passing
    the value as the single argument.

    'value_sequence' -- A sequence of values."""
    
    result = {}
    for value in value_sequence:
        result[key_function(value)] = value
    return result


def invertMap(map):
    """Produce a dictionary by inverting keys and values of 'map'."""

    return dict([ (value, key) for key, value in map.items() ])


def remove(dict, key):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    returns -- The value corresponding to 'key' in 'dict'.  

    raises -- 'KeyError' if 'dict' has no key 'key'."""
    
    value = dict[key]
    del dict[key]
    return value


def popkey(map, key, default=None):
    """Get and remove a dictionary item.

    'dict' -- The dictionary from which to get and remove the item.

    'key' -- The key of the item to get and remove.

    'default' -- The value to return if 'key' is not in 'map'.

    returns -- The value corresponding to 'key' in 'dict', or 'default'
    if none."""

    if key in map:
        result = map[key]
        del map[key]
        return result
    else:
        return default


def getHostName():
    """Return the name of this host."""

    import socket

    try:
        return socket.gethostbyname_ex(socket.gethostname())[0]
    except socket.error:
        return "(unknown)"


def isMap(object):
    """Determine whether an object satisfies the mapping protocol.

    returns -- True if 'object' satisfies the map protocol as
    specified.""" 

    # FIXME.  There probably is a more consistent way to do this.

    # If the object is an instance of the standard 'dict' class (or a
    # subclass of it), it must satisfy the mapping protocol.
    if isinstance(object, dict):
        return True
    # Check for the required methods.
    for method_name in ("get", "__getitem__", "__len__", ):
        if not hasattr(object, method_name):
            return False
    # The object checks out.
    return 1


def mkdirRecursively(path):
    """Create a directory, including parent directories if necessary."""

    parent, name = os.path.split(path)
    if parent != "" and not os.path.isdir(parent):
        mkdirRecursively(parent)
    os.mkdir(path)


def rmRecursively(path):
    """Remove the file or directory at 'path'.

    If 'path' is a directory, its contents are removed first."""

    if os.path.isdir(path):
        # Construct a list of contents of the directory.
        contents = map(lambda e, path=path: os.path.join(path, e),
                       os.listdir(path))
        # Remove them all.
        map(rmRecursively, contents)
        # Remove the directory itself.
        os.rmdir(path)
    else:
        os.unlink(path)


