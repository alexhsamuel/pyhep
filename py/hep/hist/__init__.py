#-----------------------------------------------------------------------
#
# module hep.hist
#
# Copyright 2003 by Alex Samuel.  All rights reserved.
#
#-----------------------------------------------------------------------

"""Histograms.

The following error models are supported:

  'explicit' -- Asymmetrical errors may be specified for each bin using
  the 'setError' method.  No errors are computed automatically.

  'gaussian' -- The symmetrical error on the value of each bin is
  computed assuming a Gaussian distribution.

  'none' -- Bins have zero errors.

  'poisson' -- The asymmetrical error on the value of each bin is computed
  assuming a Poisson distribution.

  'quadrature' -- The symmetrical error on each bin is square root of
  the sum of squares of weights accumulated to that bin.

Error values are always of type 'float'.
"""

#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from __future__ import generators

from   axis import Axis, BinnedAxis, AxisIterator, AxesIterator
import copy
import copy_reg
from   function import Function1D, isFunction, SampledFunction1D, \
                       sampleFunction1D
from   function import Graph, ContourGraph, isGraph, \
                       makeContourGraphFromHist
from   hep import ext
from   hep.bool import *
from   histogram import Histogram, Histogram1D, isHistogram
from   scatter import isScatter, Scatter
import sys
from   util import *

