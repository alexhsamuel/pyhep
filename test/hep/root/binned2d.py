#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

from   hep.hist import Histogram
import hep.root
from   hep.test import compare
from   math import sqrt

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

# Create and fill a histogram.
histogram = Histogram(
    (10, (0.0, 10.0)), (10, (-1.0, 1.0)), bin_type=float)
for i in xrange(0, 100):
    for j in xrange(-10, 10):
        histogram << (sqrt(i), 0.1 * j + 0.05)
histogram.accumulate((-10, -10), 42.0)
histogram.accumulate((-10, 0.1), 17.5)
histogram.accumulate((0.5, 10), -4.0)
histogram.accumulate((-10, 10), 3.33)
# Write it to an ROOT file.
hep.root.create("binned2d.root")["testhist"] = histogram
del histogram

# Load the histogram back from the file.
histogram = hep.root.open("binned2d.root")["testhist"]
# Check that it's configured correctly.
compare(histogram.dimensions, 2)
compare(histogram.bin_type, float)
x_axis, y_axis = histogram.axes
compare(x_axis.number_of_bins, 10)
compare(x_axis.range, (0.0, 10.0))
compare(x_axis.type, float)
compare(y_axis.number_of_bins, 10)
compare(y_axis.range, (-1.0, 1.0))
compare(y_axis.type, float)
# Do some checks on the contents.
compare(histogram.number_of_samples, 2004)
total = 0
for i in xrange(0, 10):
    for j in xrange(0, 10):
        total += histogram.getBinContent((i, j))
compare(total, 2000)
compare(histogram.getBinContent(("underflow", "underflow")), 42.0)
compare(histogram.getBinContent(("underflow", 5)), 17.5)
compare(histogram.getBinContent((0, "overflow")), -4.0)
compare(histogram.getBinContent((1, "overflow")), 0)
compare(histogram.getBinContent(("underflow", "overflow")), 3.33, precision=1e-6)
compare(histogram.getBinContent(("overflow", "overflow")), 0)
