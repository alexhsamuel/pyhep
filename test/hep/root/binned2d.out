Traceback (most recent call last):
  File "binned2d.py", line 25, in ?
    hep.root.create("binned2d.root")["testhist"] = histogram
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 134, in __setitem__
    subdir.__set(key, value)
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 441, in __set
    self._set(key, value, **kw_args)
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 250, in _set
    tkey = self.__getTKey(key)
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 262, in __getTKey
    return _cast(key)
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 600, in _cast
    parts = object.this.split("_")
AttributeError: 'PySwigObject' object has no attribute 'split'
